lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,296 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package desktop
+
+import (
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// DirectoryAccess allows access to the shared directory.
+// Should be kept in sync with web/packages/shared/libs/tdp/sharedDirectoryAccess.ts
+// where FS events are handled for Web UI.
+type DirectoryAccess struct {
+	// safePathGetter allows building a safe path by joining the shared directory path
+	// and a relative path.
+	// The shared directory path is not exposed to avoid unsafe operations.
+	//
+	// TODO(gzdunek): This code can be greatly simplified with os.OpenRoot.
+	// Switch to it when branch/v17 is updated to Go 1.24.
+	safePathGetter func(relativePath string) (string, error)
+}
+
+// FileOrDirInfo contains metadata about a file or a directory.
+type FileOrDirInfo struct {
+	Size         int64
+	LastModified int64
+	FileType     FileType // "file" or "directory"
+	IsEmpty      bool
+	Path         string
+}
+
+type FileType uint32
+
+const (
+	FileTypeFile FileType = iota
+	FileTypeDir
+)
+
+const StandardDirSize = 4096
+
+func (s *DirectoryAccess) Open(baseDir string) error {
+	if s.safePathGetter != nil {
+		return errors.New("only one shared directory is supported")
+	}
+
+	base, err := filepath.EvalSymlinks(baseDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(base)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !stat.IsDir() {
+		return trace.BadParameter("%q is not a directory", baseDir)
+	}
+
+	s.safePathGetter = func(relativePath string) (string, error) {
+		full := filepath.Join(base, relativePath)
+		resolved, err := filepath.EvalSymlinks(full)
+		if err != nil {
+			// EvalSymlinks returns an error if the target file does not exist.
+			// In that case, attempt to resolve the symlinks of the parent directory instead.
+			if os.IsNotExist(err) {
+				parent := filepath.Dir(full)
+				resolvedParent, perr := filepath.EvalSymlinks(parent)
+				if perr != nil {
+					return "", trace.Wrap(perr)
+				}
+
+				// Reconstruct the full path by joining the resolved parent with the original file name.
+				resolved = filepath.Join(resolvedParent, filepath.Base(full))
+			} else {
+				return "", trace.Wrap(err)
+			}
+		}
+		if !isSubPath(base, resolved) {
+			return "", trace.BadParameter("path escapes from parent")
+		}
+		return resolved, nil
+	}
+	return nil
+}
+
+func isSubPath(parent, child string) bool {
+	return child == parent || strings.HasPrefix(child, parent+string(filepath.Separator))
+}
+
+// Stat retrieves metadata about a file or directory at the given path.
+func (s *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
```

Comment:
If someone calls `Stat` prior to `Open`, `safePathGetter` will be nil and this will panic. Should we add a check for that?